kvbc/src/v4blockchain/detail/blockchain.cpp

```diff
@@ -64,6 +64,43 @@ BlockId Blockchain::addBlock(const concord::kvbc::categorization::Updates& categ
   return id;
 }
 
+// Delete up to until not including until,
+// returns the last block id that was deleted.
+BlockId Blockchain::deleteBlocksUntil(BlockId until) {
+  ConcordAssertGT(genesis_block_id_, INVALID_BLOCK_ID);
+  ConcordAssertLT(genesis_block_id_, until);
+  // We have a single block on the chain
+  if (last_reachable_block_id_ == genesis_block_id_) return genesis_block_id_ - 1;
+  // We don't want to erase all the blockchain
+  const auto last_deleted_block = std::min(last_reachable_block_id_.load() - 1, until - 1);
+  auto write_batch = native_client_->getBatch();
+  for (uint64_t i = genesis_block_id_; i <= last_deleted_block; ++i) {
+    deleteBlock(i, write_batch);
+  }
+  native_client_->write(std::move(write_batch));
+  auto blocks_deleted = (last_deleted_block - genesis_block_id_) + 1;
+  genesis_block_id_ = last_deleted_block + 1;
+
+  LOG_INFO(V4_BLOCK_LOG, "Deleted " << blocks_deleted << " blocks, new genesis is " << genesis_block_id_);
+  return last_deleted_block;
+}
+
+void Blockchain::deleteGenesisBlock() {
+  ConcordAssertGT(genesis_block_id_, INVALID_BLOCK_ID);
+  ConcordAssertLT(genesis_block_id_, last_reachable_block_id_);
+  auto write_batch = native_client_->getBatch();
+  deleteBlock(genesis_block_id_, write_batch);
+  native_client_->write(std::move(write_batch));
+  ++genesis_block_id_;
+  LOG_INFO(V4_BLOCK_LOG, "Deleted genesis, new genesis is " << genesis_block_id_);
+}
+
+void Blockchain::deleteLastReachableBlock(storage::rocksdb::NativeWriteBatch& write_batch) {
```

Comment:
Should we change this function to:
void Blockchain::deleteLastReachableBlock();
Just like other deletes it will create its batch and call the deleteBlock.

Reply:
since the delete last reachable requires deleting also from the latest CF , it needs to be an atomic operation.
Therefore passing the write_batch to here and then to the latest to add its updates is required.